magda-authorization-api/src/getWhoAllowDatasetOperation.ts

```diff
@@ -0,0 +1,199 @@
+        }
+        return `(SELECT Children."id"
+        FROM org_units AS Parents, org_units AS Children
+        WHERE Children."left" >= Parents."left" AND Children."left" <= Parents."right" AND Parents."id" = users."orgUnitId")`;
+    },
+    "input.user.id": "users.id",
+    "input.user.permissions[_].resourceUri": "resources.uri"
+};
+
+function isValidRef(ref: string): boolean {
+    return (
+        Object.keys(VALID_REF_PREFIX_LIST).findIndex(
+            item => ref.indexOf(item) === 0
+        ) !== -1
+    );
+}
+
+function regoRefToSqlIdentifier(ref: RegoRef): string {
+    const refString = ref.fullRefString();
+    if (!isValidRef(refString)) {
+        throw new Error(`Invalid ref: ${ref.fullRefString()}`);
+    }
+    const prefixes = Object.keys(VALID_REF_PREFIX_LIST).sort(
+        (a: string, b: string) => b.length - a.length
+    );
+    for (let i = 0; i < prefixes.length; i++) {
+        const prefix = prefixes[i];
+        if (!prefix) continue;
+        const idx = refString.indexOf(prefix);
+        if (idx !== 0) continue;
+        if (typeof VALID_REF_PREFIX_LIST[prefix] === "function") {
+            return (VALID_REF_PREFIX_LIST[prefix] as Function)(refString);
+        } else if (typeof VALID_REF_PREFIX_LIST[prefix] === "string") {
+            return VALID_REF_PREFIX_LIST[prefix] as string;
+        } else {
+            throw new Error("Invalid VALID_REF_PREFIX_LIST item type!");
+        }
+    }
+    return refString;
+}
+
+function regoExpToSql(exp: RegoExp, sqlParameters: any[]): string {
+    const [operator, operands] = exp.toOperatorOperandsArray();
+    if (operands.findIndex(term => !term.isResolveAsCollectionValue()) === -1) {
+        // --- both operands are CollectionRef? not possible
+        throw new Error(`Invalid expression ${exp.toHumanReadableString()}`);
+    }
+    let operatorString = operator;
+    if (operands.findIndex(term => term.isResolveAsCollectionValue()) !== -1) {
+        // --- one ref can be resolved to a collection value
+        // --- convert operator to relevant collection operators
+        if (operatorString === "=") {
+            operatorString = "IN";
+        } else if (operatorString === "!=") {
+            operatorString = "NOT IN";
+        } else {
+            throw new Error(
+                `Invalid operator found for expression: ${exp.toHumanReadableString()}`
+            );
+        }
+    }
+    const operandParts = operands.map(operandTerm => {
+        if (operandTerm.isRef()) {
+            return regoRefToSqlIdentifier(operandTerm.getRef());
+        } else {
+            const value = operandTerm.getValue();
+            sqlParameters.push(value);
+            return isUuid(value)
+                ? `$${sqlParameters.length}::uuid`
+                : `$${sqlParameters.length}`;
+        }
+    });
+    return `${operandParts[0]} ${operatorString} ${operandParts[1]}`;
+}
+
+async function getWhoAllowDatasetOperation(
```

Comment:
`getWhoAllowDatasetOperation` is a really awkward phrase grammar-wise - could it be `getUsersAllowedOperationOnDataset` maybe? 🤔

Reply:
Good name~ 👍

Reply:
changed